packages/plugin-ext/src/plugin/languages/reference.ts

```diff
@@ -0,0 +1,63 @@
+/********************************************************************************
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import URI from 'vscode-uri/lib/umd';
+import * as theia from '@theia/plugin';
+import { DocumentsExtImpl } from '../documents';
+import * as types from '../types-impl';
+import { ReferenceContext, Location } from '../../api/model';
+import * as Converter from '../type-converters';
+import { Position } from '../../api/plugin-api';
+import { createToken } from '../token-provider';
+
+export class ReferenceAdapter {
+
+    constructor(
+        private readonly provider: theia.ReferenceProvider,
+        private readonly documents: DocumentsExtImpl
+    ) { }
+
+    provideReferences(resource: URI, position: Position, context: ReferenceContext): Promise<Location[] | undefined> {
+        const documentData = this.documents.getDocumentData(resource);
+        if (!documentData) {
+            return Promise.reject(new Error(`There is no document for ${resource}`));
+        }
+
+        const document = documentData.document;
+        const zeroBasedPosition = Converter.toPosition(position);
+
+        return Promise.resolve(this.provider.provideReferences(document, zeroBasedPosition, context, createToken())).then(referencce => {
+            if (!referencce) {
+                return undefined;
+            }
+
+            if (isLocationArray(referencce)) {
+                const locations: Location[] = [];
+
+                for (const location of referencce) {
+                    locations.push(Converter.fromLocation(location));
+                }
+
+                return locations;
+            }
+        });
+    }
+}
+
+/* tslint:disable-next-line:no-any */
+function isLocationArray(array: any): array is types.Location[] {
```

Comment:
why is it not a private method ? also do we have type of reference so we could avoid any ?

Reply:
`isLocationArray` is not exported, so it's visible only inside its module, I can't put `private` modifier on it (unless I make it a member of `ReferenceAdapter` class). Honestly, it's `copy-pasted` from `./definition.ts`, but I think, as it's simple and duplicated only once,  it's reasonable to leave it as it is, or, if needed later, to move such functions into some `utility.ts`-like module and make them exported.

According to https://github.com/Microsoft/vscode/blob/5a4eab812bd485942061f71af39f3eabc52a1035/src/vs/vscode.d.ts#L2604 i don't see any kind of `types` for the references.

Reply:
if it's only used internally, then I still don't see why not use it as private method of ReferenceAdapter ?

Reply:
I've added a fixup for this.